text/0000-official-utility-buildpacks.md

```diff
@@ -0,0 +1,75 @@
+# Meta
+[meta]: #meta
+- Name: Officially supported utility buildpacks
+- Start Date: 2021-06-30
+- Author(s): [@jkutner](https://github.com/jkutner)
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+This is a proposal to adopt utility buildpacks as officially supported components of the Cloud Native Buildpacks project.
+
+# Definitions
+[definitions]: #definitions
+
+- **utility buildpack** - a buildpack that provides a generic capability that is useful across many vendors, platforms, and apps
+- **officially supported** - supported and maintained by the the Cloud Native Buildpacks project and its members
+
+# Motivation
+[motivation]: #motivation
+
+Many capabilities of lifecycle and pack do not require a native implementation within these components. These capabilities can (and have been) implemented at the user layer as buildpacks (ex. [inline-buildpack](https://registry.buildpacks.io/buildpacks/hone/inline)). In [PR #168](https://github.com/buildpacks/rfcs/pull/168) we assert that supporting these capabilities natively adds complexity to the implementation of lifecycle, which makes it difficult to innovate and improve.
+
+However, defering to the community and vendors to implement these kinds of utility buildpacks has not necessarily had good results. For example, there are three implementations of the [Procfile buildpack](https://registry.buildpacks.io/buildpacks/paketo-buildpacks/procfile). None of these implementations add significant (if any) advantages over the other. Yet buildpack users are force to decide which one to select. This is an unneccessary burden for a capability that could be provided by default.
+
+We seek to reduce complexity in the lifecycle and provide well maintained utility buildpacks for users by supporting them as official components of the Cloud Native Buildpacks project. This also aligns with our "everything's a buildpack" mantra.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Reserve the `cnb` namespace for officially supported buildpacks. Then, on a case by case basis, we adopt utility buildpacks under this
```

Comment:
I'm a little worried that this opens a bit of a pandora's box.

We might have a clear distinction of utility buildpacks and _other buildpacks_. But newcomers might likely just be seeing **"officially supported buildpacks"**. Once we start doing this, I'm worried that we'll be pressured to have more of these officially supported buildpacks. If it were me, I'd associate the `cnb/` namespace with the `library/` namespace on DockerHub (_which is what I think cnb namespace should be called, btw_).

I'm sure that newcomers are a bit thrown when they find that they have to go to the [paketo](https://paketo.io/) project to find all of their buildpacks. Which leads me to worry that we will be pit in a (_losing_) competition with the paketo project. If it's going be ours vs theirs, we'll have a hard time covering the ground that they already have.

And finally, "owning, supporting, and maintaining" and "testing buildpacks across stacks" _seems like_ a lot of responsibilities on top of what Implementation team probably already has to do.

But I like the idea 🙂

Reply:
+1 to the Pandora's box comment. I think that having some "official" buildpacks will absolutely prompt people to request more. So where do you draw the line? Technically you only need one buildpack that provides a language runtime binary, should that become an official buildpack too?

In regards to newcomers, I think that can be mitigated with good docs & collaboration with the overall ecosystem. The Buildpacks projects can and should direct users to community resources (in a fair, taking no favorites way). Inform users that there are choices and allow them to pick what makes the most sense for their needs.

Reply:
These are a very real concerns, and they are the reasons we never included buildpacks in the project in the past.

What makes the buildpacks in this proposal a little different is that they _replace_ native capabilities. In some sense, the support burden is equal to the existing support burden.

I'm not worried about being pitted against Paketo. They should be a different set of buildpacks. If Paketo has utility buildpacks that need to be contributed upstream we can consider accepting them.

Reply:
>What makes the buildpacks in this proposal a little different is that they replace native capabilities. In some sense, the support burden is equal to the existing support burden.

Isn't the idea to reduce burden though? If the burden is equal, then doesn't that make a case for implementing these features in the lifecycle/pack?

IMHO, you can reduce the burden by making it clear that those features won't be part of the buildpack and allowing the community time to fill in the features through buildpacks (see my other comment below).

Reply:
Our goal is to reduce burden by changing the architecture (i.e. moving logic out of the lifecycle, and reducing coupling). I do not have a goal of removing capabilities/features from the project.

Reply:
> I do not have a goal of removing capabilities/features from the project.

Not suggesting that specifically. You mentioned Procfile in the RFC, so that is what I'm thinking of here. There's no need for buildpacks to directly have a Procfile buildpack because it was a need identified by the downstream communities and they have provided support for it.

All I'm saying is that if there's a clear delineation between what the buildpacks will and won't do, that provides guides and leaves room for the community to come in and implement any features that are marked out of scope but are important to the downstream communities.

Reply:
If we're calling them system buildpacks maybe we can use `cnb-system/` as the namespace?